porchlight/neighborhood.py

```diff
@@ -426,6 +457,213 @@ def call_all_doors(self):
+        positional-only arguments and/or no no keyword arguments, then empty
+        objects are returned.
+
+        Arguments
+        ---------
+        input_door : :py:class:`~porchlight.door.Door`
+            The door to gather necessary parameters for.
+
+        Returns
+        -------
+        args : list
+            Positional-only arguments required by door.
+
+        kwargs : dict[str, Any]
+            Keyword arguments for the door.
+
+        Notes
+        -----
+        The return values must be unpacked before being used to call the
+        :py:class:`~porchlight.door.Door`.
+        """
+        # Gather the arguments needed by the door.
+        req_params = input_door.arguments
+        input_params = {p: self._params[p].value for p in req_params}
+
+        args = []
+        kwargs = {}
+
+        for p, value in input_params.items():
+            if p in input_door.positional_only:
+                args.append(value)
+
+            else:
+                kwargs[p] = value
+
+        return args, kwargs
+
+    def initialize(self):
+        """Runs initialization functions present in
+        :py:attr:`~porchlight.neighborhood.Neighborhood.initialization` if
+        ``has_initialized`` is ``False`` for this
+        :py:class:`~porchlight.neighborhood.Neighborhood`.
+        """
+        # Do nothing if initialization has already happened.
+        if self.has_initialized or not self.initialization:
+            return
+
+        # Ensure initialization is iterable, if not
+        if not hasattr(self.initialization, "__iter__"):
+            self.initialization = [self.initialization]
+
+            # Log this any time initialization happens, because this is
+            # important.
+            logging.info(
+                f"Could not iterate through object: "
+                f"{type(self.initialization)}, will now attempt to call "
+                f"the object."
+            )
+
+        # Ensure that none of the doors are DynamicDoors, as this can cause
+        # unexpected behaviour.
+        if any(isinstance(f, door.DynamicDoor) for f in self.initialization):
+            msg = "DynamicDoors are not yet supported for initialization."
+
+            logging.error(msg)
+            raise NotImplementedError(msg)
+
+        # Call each of the callables in the initialization function. If they
+        # are doors, save their parameter states.
+        for fxn in self.initialization:
+            if not isinstance(fxn, door.Door):
+                fxn = door.Door(fxn)
+
+            return_values = fxn.return_vals
+            arguments, keyword_arguments = self.gather_door_arguments(fxn)
+
+            result = fxn(*arguments, **keyword_arguments)
+
+            # Update any parameters already present. For initialization, only
+            # recognized return values are updated.
+            #
+            # TODO: ^ This behaviour needs to be better documented.
```

Comment:
Document this

Reply:
Done in commit https://github.com/teald/porchlight/pull/66/commits/649365797e12df234f8a4e872f7996dcb16b6d6f